File: dfs_bfs_backtracking/cpp/b1058.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

void bfs(std::vector<std::vector<int>>& connections, const int& N, int& max_num) {
	for (int person = 0; person < N; ++person) {
		int temp_max_num = 0;
		int visited[50]{ 0 };

		std::queue<int> check_queue;
		check_queue.push(person);
		visited[person] = 1;

		while (check_queue.size()) {
			int check_person = check_queue.front();
			check_queue.pop();

			if (visited[check_person] > 1) {
				temp_max_num += 1;
				if (visited[check_person] == 3) {
					continue;
				}
			}

			for (int& connection : connections[check_person]) {
				if (!visited[connection]) {
					check_queue.push(connection);
					visited[connection] = visited[check_person] + 1;
				}
			}
		}

		if (temp_max_num > max_num) {
			max_num = temp_max_num;
		}
	}
}

int main() {
	int N;
	std::cin >> N;
	std::vector<std::vector<int>> connections;
	connections.resize(N);
	for (int i = 0; i < N; ++i) {
		std::string temp_connections;
		std::cin >> temp_connections;
		for (int idx = 0; idx < temp_connections.size(); ++idx) {
			if (temp_connections[idx] == 'Y') {
				connections[i].push_back(idx);
			}
		}
	}

	int max_num = 0;
	bfs(connections, N, max_num);
	std::cout << max_num << std::endl;
}

File: dfs_bfs_backtracking/python/b1058.py
def bfs(connections, N):
    max_num = 0
    for person in range(N):
        temp_num = 0
        visited = [0] * N
        check_queue = [0] * N
        front = 0
        rear = 1 
        check_queue[0] = person
        visited[person] = 1
        while front < rear:
            check = check_queue[front]
            front += 1
            if visited[check] > 1:
                temp_num += 1
                if visited[check] == 3:
                    continue
            for connection in connections[check]:
                if not visited[connection]:
                    check_queue[rear] = connection
                    rear += 1
                    visited[connection] = visited[check] + 1
        if temp_num > max_num:
            max_num = temp_num
    return max_num

N = int(input())
connections = [[] for _ in range(N)]

for i in range(N):
    temp_list = input()
    for idx in range(N):
        if temp_list[idx] == 'Y':
            connections[i].append(idx)
print(bfs(connections, N))
